# Wrangle & Analyze WeRateDogs Data

<hr>

#### Table of Contents:
* [Project Description](#Project-Description)  
* [Notebook Setup](#Notebook-Setup)  
<br/> 
* [Gathering Data](#Gathering-Data)
    * [Enhanced Twitter Archive](#Gether:-Enhanced-Twitter-Archive)  
    * [Image Predictions File](#Gether:-Image-Predictions-File)  
    * [Twitter API File](#Gether:-Twitter-API-File)  
<br/> 
* [Assessing Data](#Assessing-Data)
    * [Twitter Archive Data](#Assess:-Twitter-Archive-Data)  
    * [Image Predictions](#Assess:-Image-Predictions)  
    * [Twitter API Data](#Assess:-Twitter-API-Data)  
<br/> 
* [Cleaning Data](#Cleaning-Data)
    * [Twitter Archive Data](#Clean:-Twitter-Archive-Data)  
    * [Image Predictions](#Clean:-Image-Predictions)  
    * [Twitter API Data](#Clean:-Twitter-API-Data)  
    * [Merge Datasets](#Merge-Datasets)  
<br/> 
* [Analyzing Data](#Analyzing-Data)

<hr>

## Project Description

Real-world data rarely comes clean. Using Python and its libraries, I will gather data from a variety of sources and in a variety of formats, assess its quality and tidiness, then clean it. This is called data wrangling. I will document my wrangling efforts in a Jupyter Notebook, plus showcase them through analyses and visualizations using Python.

The dataset that I will be wrangling (and analyzing and visualizing) is the tweet archive of Twitter user @dog_rates, also known as WeRateDogs. WeRateDogs is a Twitter account that rates people's dogs with a humorous comment about the dog. These ratings almost always have a denominator of 10. The numerators, though? Almost always greater than 10. 11/10, 12/10, 13/10, etc. Why? Because "they're good dogs Brent." WeRateDogs has over 4 million followers and has received international media coverage.

<p align="center">
  <img src="img/dog-rates-social.jpg" width="600">
</p>

The WeRateDogs Twitter archive contains basic tweet data for all 5000+ of their tweets, but not everything. One column the archive does contain though: each tweet's text, which I used to extract rating, dog name, and dog "stage" (i.e. doggo, floofer, pupper, and puppo) to make this Twitter archive "enhanced." Of the 5000+ tweets, I have filtered for tweets with ratings only (there are 2356).

<p align="center">
  <img src="img/data.png" width="1300">
</p>

Retweet count and favorite count are two of the notable column omissions. Fortunately, this additional data can be gathered from Twitter's API, which I will do.

## Notebook Setup

Load libraries and set pandas display options.

<hr>

In [1]:
# import libraries
import numpy as np
import pandas as pd
import requests
import json
import matplotlib.pyplot as plt
from stop_words import get_stop_words
import re
from functools import reduce

# pandas settings
pd.set_option('display.max_colwidth', -1)

## Gathering Data

Gather data from various sources and a variety of file formats.

<hr>

* [Enhanced Twitter Archive](#Gether:-Enhanced-Twitter-Archive)  
* [Image Predictions File](#Gether:-Image-Predictions-File)  
* [Twitter API File](#Gether:-Twitter-API-File)


### Gather: Enhanced Twitter Archive

This archive contains basic tweet data (tweet ID, timestamp, text, etc.) for all 5000+ of their tweets as they stood on August 1, 2017.

In [2]:
# load twitter archive
twitter_arch = pd.read_csv("data/twitter-archive-enhanced.csv")
# use tweet id column as index
twitter_arch.set_index("tweet_id", inplace = True)
# display few lines
twitter_arch.head(3)

in_reply_to_status_id  in_reply_to_user_id  \
tweet_id                                                         
892420643555336193 NaN                    NaN                    
892177421306343426 NaN                    NaN                    
891815181378084864 NaN                    NaN                    

                                    timestamp  \
tweet_id                                        
892420643555336193  2017-08-01 16:23:56 +0000   
892177421306343426  2017-08-01 00:17:27 +0000   
891815181378084864  2017-07-31 00:18:03 +0000   

                                                                                                source  \
tweet_id                                                                                                 
892420643555336193  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
892177421306343426  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
891815181378084864  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                                                  text  \
tweet_id                                                                                                                                                                                 
892420643555336193  This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU                                                        
892177421306343426  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   
891815181378084864  This is Archie. He is a rare Norwegian Pouncing Corgo. Lives in the tall grass. You never know when one may strike. 12/10 https://t.co/wUnZnhtVJB                    

                    retweeted_status_id  retweeted_status_user_id  \
tweet_id                                                            
892420643555336193 NaN                  NaN                         
892177421306343426 NaN                  NaN                         
891815181378084864 NaN                  NaN                         

                   retweeted_status_timestamp  \
tweet_id                                        
892420643555336193  NaN                         
892177421306343426  NaN                         
891815181378084864  NaN                         

                                                                      expanded_urls  \
tweet_id                                                                              
892420643555336193  https://twitter.com/dog_rates/status/892420643555336193/photo/1   
892177421306343426  https://twitter.com/dog_rates/status/892177421306343426/photo/1   
891815181378084864  https://twitter.com/dog_rates/status/891815181378084864/photo/1   

                    rating_numerator  rating_denominator     name doggo  \
tweet_id                                                                  
892420643555336193  13                10                  Phineas  None   
892177421306343426  13                10                  Tilly    None   
891815181378084864  12                10                  Archie   None   

                   floofer pupper puppo  
tweet_id                                 
892420643555336193  None    None   None  
892177421306343426  None    None   None  
891815181378084864  None    None   None

### Gether: Image Predictions File

This file contains top three predictions of dog breed for each dog image from the WeRateDogs archive. Table contains the top three predictions, tweet ID, image URL, and the image number that corresponded to the most confident prediction (numbered 1 to 4 since tweets can have up to four images).

In [3]:
# get file with the image predictions
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
with open('data/image-predictions.tsv' , 'wb') as file:
    predictions = requests.get(url)
    file.write(predictions.content)

# load image predictions
image_pred = pd.read_csv('data/image-predictions.tsv', sep = '\t')
# use tweet id column as index
image_pred.set_index("tweet_id", inplace = True)
# display few lines
image_pred.head(3)

jpg_url  img_num  \
tweet_id                                                                       
666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg  1         
666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg  1         
666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg  1         

                                        p1   p1_conf  p1_dog  \
tweet_id                                                       
666020888022790149  Welsh_springer_spaniel  0.465074  True     
666029285002620928  redbone                 0.506826  True     
666033412701032449  German_shepherd         0.596461  True     

                                    p2   p2_conf  p2_dog                   p3  \
tweet_id                                                                        
666020888022790149  collie              0.156665  True    Shetland_sheepdog     
666029285002620928  miniature_pinscher  0.074192  True    Rhodesian_ridgeback   
666033412701032449  malinois            0.138584  True    bloodhound            

                     p3_conf  p3_dog  
tweet_id                              
666020888022790149  0.061428  True    
666029285002620928  0.072010  True    
666033412701032449  0.116197  True

### Gether: Twitter API File

Retweet count and favorite count are two of the notable column omissions of Twitter data archive. Fortunately, this additional data can be gathered from Twitter's API. Twitter API file contains tweet id, favorite count and retweet count. 

In [4]:
# load twitter API data
with open('data/tweet-json.txt') as f:
    twitter_api = pd.DataFrame((json.loads(line) for line in f), columns = ['id', 'favorite_count', 'retweet_count'])

# change column names
twitter_api.columns = ['tweet_id', 'favorites', 'retweets']
# use tweet id column as index
twitter_api.set_index('tweet_id', inplace = True)
# display few lines
twitter_api.head(3)

favorites  retweets
tweet_id                               
892420643555336193  39467      8853    
892177421306343426  33819      6514    
891815181378084864  25461      4328


## Assessing Data

Assess data visually and programmatically for quality and tidiness issues using pandas.

<hr>

* [Twitter Archive Data](#Assess:-Twitter-Archive-Data)  
* [Image Predictions](#Assess:-Image-Predictions)  
* [Twitter API Data](#Assess:-Twitter-API-Data)

### Assess: Twitter Archive Data

In [5]:
# display sample of data
twitter_arch.sample(3)

in_reply_to_status_id  in_reply_to_user_id  \
tweet_id                                                         
682788441537560576 NaN                    NaN                    
690959652130045952 NaN                    NaN                    
667182792070062081 NaN                    NaN                    

                                    timestamp  \
tweet_id                                        
682788441537560576  2016-01-01 05:00:24 +0000   
690959652130045952  2016-01-23 18:09:53 +0000   
667182792070062081  2015-11-19 03:29:07 +0000   

                                                                                                source  \
tweet_id                                                                                                 
682788441537560576  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
690959652130045952  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
667182792070062081  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                           text  \
tweet_id                                                                                                                                                          
682788441537560576  Happy New Year from your fav holiday squad! 🎉 12/10 for all\n\nHere's to a pupper-filled year 🍻🐶🐶🐶 https://t.co/ZSdEj59FGf                    
690959652130045952  This golden is happy to refute the soft mouth egg test. Not a fan of sweeping generalizations. 11/10 #notallpuppers https://t.co/DgXYBDMM3E   
667182792070062081  This is Timison. He just told an awful joke but is still hanging on to the hope that you'll laugh with him. 10/10 https://t.co/s2yYuHabWl     

                    retweeted_status_id  retweeted_status_user_id  \
tweet_id                                                            
682788441537560576 NaN                  NaN                         
690959652130045952 NaN                  NaN                         
667182792070062081 NaN                  NaN                         

                   retweeted_status_timestamp  \
tweet_id                                        
682788441537560576  NaN                         
690959652130045952  NaN                         
667182792070062081  NaN                         

                                                                                                                                                                                                                                                                      expanded_urls  \
tweet_id                                                                                                                                                                                                                                                                              
682788441537560576  https://twitter.com/dog_rates/status/682788441537560576/photo/1                                                                                                                                                                                                   
690959652130045952  https://twitter.com/dog_rates/status/690959652130045952/photo/1,https://twitter.com/dog_rates/status/690959652130045952/photo/1,https://twitter.com/dog_rates/status/690959652130045952/photo/1,https://twitter.com/dog_rates/status/690959652130045952/photo/1   
667182792070062081  https://twitter.com/dog_rates/status/667182792070062081/photo/1                                                                                                                                                                                                   

                    rating_numerator  rating_denominator     name doggo  \
tweet_id                         

In [6]:
# print a summary of a DataFrame
twitter_arch.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2356 entries, 892420643555336193 to 666020888022790149
Data columns (total 16 columns):
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(2), object(10)
memory usa

In [7]:
# check if ids are unique
twitter_arch.index.is_unique

True

In [8]:
# check number of replies
np.isfinite(twitter_arch.in_reply_to_status_id).sum()

78

In [9]:
# check values in sources
twitter_arch.source.value_counts()

<a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>     2221
<a href="http://vine.co" rel="nofollow">Vine - Make a Scene</a>                        91  
<a href="http://twitter.com" rel="nofollow">Twitter Web Client</a>                     33  
<a href="https://about.twitter.com/products/tweetdeck" rel="nofollow">TweetDeck</a>    11  
Name: source, dtype: int64

In [10]:
# check quality of text
twitter_arch.text.sample(3)

tweet_id
739238157791694849    Here's a doggo blowing bubbles. It's downright legendary. 13/10 would watch on repeat forever (vid by Kent Duryee) https://t.co/YcXgHfp1EC
719704490224398336    This is Clyde. He's making sure you're having a good train ride. 12/10 great pupper https://t.co/y206kWHAj0                               
680221482581123072    This is CeCe. She's patiently waiting for Santa. 10/10 https://t.co/ZJUypFFwvg                                                            
Name: text, dtype: object

In [11]:
# check number of retweets
np.isfinite(twitter_arch.retweeted_status_id).sum()

181

In [12]:
# check expanded urls
twitter_arch[~twitter_arch.expanded_urls.str.startswith(('https://twitter.com','http://twitter.com', 'https://vine.co'), na=False)].sample(3)[['text','expanded_urls']]

text  \
tweet_id                                                                                                                                                             
747651430853525504  Other pupper asked not to have his identity shared. Probably just embarrassed about the headbutt. Also 12/10 it'll be ok mystery pup             
674330906434379776  13/10\n@ABC7                                                                                                                                     
868639477480148993  RT @dog_rates: Say hello to Cooper. His expression is the same wet or dry. Absolute 12/10 but Coop desperately requests your help\n\nhttps://…   

                                                                                                                                                                       expanded_urls  
tweet_id                                                                                                                                                                              
747651430853525504  NaN                                                                                                                                                               
674330906434379776  NaN                                                                                                                                                               
868639477480148993  https://www.gofundme.com/3ti3nps,https://twitter.com/dog_rates/status/868552278524837888/photo/1,https://twitter.com/dog_rates/status/868552278524837888/photo/1

In [13]:
# check for two or more urls in the expanded urls
twitter_arch[twitter_arch.expanded_urls.str.contains(',', na = False)].expanded_urls.count()

639

In [14]:
# check rating denominator
twitter_arch.rating_denominator.value_counts()

10     2333
11     3   
50     3   
80     2   
20     2   
2      1   
16     1   
40     1   
70     1   
15     1   
90     1   
110    1   
120    1   
130    1   
150    1   
170    1   
7      1   
0      1   
Name: rating_denominator, dtype: int64

In [15]:
# check ratings with denominator greather than 10
twitter_arch[twitter_arch.rating_denominator > 10][['text', 'rating_denominator']]

text  \
tweet_id                                                                                                                                                                 
832088576586297345  @docmisterio account started on 11/15/15                                                                                                             
820690176645140481  The floofs have been released I repeat the floofs have been released. 84/70 https://t.co/NIYC820tmd                                                  
775096608509886464  RT @dog_rates: After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https:/…         
758467244762497024  Why does this never happen at my front door... 165/150 https://t.co/HmwrdfEfUE                                                                       
740373189193256964  After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https://t.co/XAVDNDaVgQ         
731156023742988288  Say hello to this unbelievably well behaved squad of doggos. 204/170 would try to pet all at once https://t.co/yGQI3He3xv                            
722974582966214656  Happy 4/20 from the squad! 13/10 for all https://t.co/eV1diwds8a                                                                                     
716439118184652801  This is Bluebert. He just saw that both #FinalFur match ups are split 50/50. Amazed af. 11/10 https://t.co/Kky1DPG4iq                                
713900603437621249  Happy Saturday here's 9 puppers on a bench. 99/90 good work everybody https://t.co/mpvaVxKmc1                                                        
710658690886586372  Here's a brigade of puppers. All look very prepared for whatever happens next. 80/80 https://t.co/0eb7R1Om12                                         
709198395643068416  From left to right:\nCletus, Jerome, Alejandro, Burp, &amp; Titson\nNone know where camera is. 45/50 would hug all at once https://t.co/sedre1ivTK   
704054845121142784  Here is a whole flock of puppers.  60/50 I'll take the lot https://t.co/9dpcw6MdWa                                                                   
697463031882764288  Happy Wednesday here's a bucket of pups. 44/40 would pet all at once https://t.co/HppvrYuamZ                                                         
686035780142297088  Yes I do realize a rating of 4/20 would've been fitting. However, it would be unjust to give these cooperative pups that low of a rating             
684225744407494656  Two sneaky puppers were not initially seen, moving the rating to 143/130. Please forgive us. Thank you https://t.co/kRK51Y5ac3                       
684222868335505415  Someone help the girl is being mugged. Several are distracting her while two steal her shoes. Clever puppers 121/110 https://t.co/1zfnTJLt55         
682962037429899265  This is Darrel. He just robbed a 7/11 and is in a high speed police chase. Was just spotted by the helicopter 10/10 https://t.co/7EsP8LmSp5          
682808988178739200  I'm aware that I could've said 20/16, but here at WeRateDogs we are very professional. An inconsistent rating scale is simply irresponsible          
677716515794329600  IT'S PUPPERGEDDON. Total of 144/120 ...I think https://t.co/ZanVtAtvIq                                                                               
675853064436391936  Here we have an entire platoon of puppers. Total score: 88/80 would pet all at once https://t.co/y93p6FLvVw                                          

                    rating_denominator  
tweet_id                                
832088576586297345  15                  
820690176645140481  70                  
775096608509886464  11                  
758467244762497024  150                 
740373189193256964  11                  
731156023742988288  170                 
722974582966214656  20                  
716439118184652801  50                  
71390060343

In [16]:
# check rating numerator
twitter_arch.rating_numerator.value_counts()

12      558
11      464
10      461
13      351
9       158
8       102
7       55 
14      54 
5       37 
6       32 
3       19 
4       17 
1       9  
2       9  
420     2  
0       2  
15      2  
75      2  
80      1  
20      1  
24      1  
26      1  
44      1  
50      1  
60      1  
165     1  
84      1  
88      1  
144     1  
182     1  
143     1  
666     1  
960     1  
1776    1  
17      1  
27      1  
45      1  
99      1  
121     1  
204     1  
Name: rating_numerator, dtype: int64

In [17]:
# check for any float ratings in the text column
twitter_arch[twitter_arch.text.str.contains(r'\d+\.\d+\/\d+')][['text','rating_denominator', 'rating_numerator']]

text  \
tweet_id                                                                                                                                                                                
883482846933004288  This is Bella. She hopes her smile made you smile. If not, she is also offering you her favorite monkey. 13.5/10 https://t.co/qjrljjt948                            
832215909146226688  RT @dog_rates: This is Logan, the Chow who lived. He solemnly swears he's up to lots of good. H*ckin magical af 9.75/10 https://t.co/yBO5wu…                        
786709082849828864  This is Logan, the Chow who lived. He solemnly swears he's up to lots of good. H*ckin magical af 9.75/10 https://t.co/yBO5wuqaPS                                    
778027034220126208  This is Sophie. She's a Jubilant Bush Pupper. Super h*ckin rare. Appears at random just to smile at the locals. 11.27/10 would smile back https://t.co/QFaUiIHxHq   
681340665377193984  I've been told there's a slight possibility he's checking his mirror. We'll bump to 9.5/10. Still a menace                                                          
680494726643068929  Here we have uncovered an entire battalion of holiday puppers. Average of 11.26/10 https://t.co/eNm2S6p9BD                                                          

                    rating_denominator  rating_numerator  
tweet_id                                                  
883482846933004288  10                  5                 
832215909146226688  10                  75                
786709082849828864  10                  75                
778027034220126208  10                  27                
681340665377193984  10                  5                 
680494726643068929  10                  26

In [18]:
# check name of dog
twitter_arch.name.value_counts()

None         745
a            55 
Charlie      12 
Cooper       11 
Oliver       11 
Lucy         11 
Penny        10 
Tucker       10 
Lola         10 
Bo           9  
Winston      9  
the          8  
Sadie        8  
Daisy        7  
Bailey       7  
an           7  
Toby         7  
Buddy        7  
Stanley      6  
Leo          6  
Koda         6  
Bella        6  
Jack         6  
Scout        6  
Oscar        6  
Dave         6  
Milo         6  
Jax          6  
Rusty        6  
George       5  
            ..  
Halo         1  
Timofy       1  
Tango        1  
Nigel        1  
Stephanus    1  
Danny        1  
Sweet        1  
Amélie       1  
by           1  
Jockson      1  
Mabel        1  
Carper       1  
Angel        1  
Blanket      1  
Beckham      1  
Eazy         1  
Rover        1  
Stefan       1  
Olaf         1  
Deacon       1  
Karl         1  
William      1  
Hanz         1  
Andru        1  
Tommy        1  
Grady        1  
Julius       1  
Fiji         1

In [19]:
# check for stop words in dog name
# https://stackoverflow.com/a/5486535/7382214
stop_words = set(get_stop_words('en'))

count = 0
for word in twitter_arch.name:
    if word.lower() in stop_words:
        count += 1
print('Rows with stop words:', count)

Rows with stop words: 83


In [20]:
# check if dogs have more than one category assigned
categories = ['doggo', 'floofer', 'pupper', 'puppo']

for category in categories:
    twitter_arch[category] = twitter_arch[category].apply(lambda x: 0 if x == 'None' else 1)

twitter_arch['number_categories'] = twitter_arch.loc[:,categories].sum(axis = 1)

In [21]:
# dogs categories
twitter_arch['number_categories'].value_counts()

0    1976
1    366 
2    14  
Name: number_categories, dtype: int64

#### Quality & Tidiness Issues

- some of the gathered tweets are replies and should be removed;
- the timestamp has an incorrect datatype - is an object, should be DateTime;
- source is an HTML element - its text should be extracted;
- some rows in the text column begin from 'RT @dog_rates:';
- some rows in the text column have leading and/or trailing whitespace;
- some of the gathered tweets are retweets;
- we have 59 missing expanded urls;
- we have 639 expanded urls which contain more than one url address;
- denominator of some ratings is not 10;
- numerator of some ratings is greater than 10;
- float ratings have been incorrectly read from the text of tweet;
- 'None' in the name should be convert to NaN;
- we have stop words in the name column;
- dog 'stage' classification (doggo, floofer, pupper or puppo) should be one column;
- some dogs have more than one category assigned;

### Assess: Image Predictions

In [22]:
# display sample of data
image_pred.sample(3)

jpg_url  img_num  \
tweet_id                                                                       
842846295480000512  https://pbs.twimg.com/media/C7JkO0rX0AErh7X.jpg  1         
808838249661788160  https://pbs.twimg.com/media/CzmSFlKUAAAQOjP.jpg  1         
694669722378485760  https://pbs.twimg.com/media/CaP2bS8WYAAsMdx.jpg  2         

                                    p1   p1_conf  p1_dog                  p2  \
tweet_id                                                                       
842846295480000512  Labrador_retriever  0.461076  True    golden_retriever     
808838249661788160  Rottweiler          0.369530  True    miniature_pinscher   
694669722378485760  beaver              0.457094  False   mongoose             

                     p2_conf  p2_dog         p3   p3_conf  p3_dog  
tweet_id                                                           
842846295480000512  0.154946  True    Chihuahua  0.110249  True    
808838249661788160  0.194867  True    kelpie     0.160104  True    
694669722378485760  0.228298  False   marmot     0.148309  False

In [23]:
# print a summary of a DataFrame
image_pred.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2075 entries, 666020888022790149 to 892420643555336193
Data columns (total 11 columns):
jpg_url    2075 non-null object
img_num    2075 non-null int64
p1         2075 non-null object
p1_conf    2075 non-null float64
p1_dog     2075 non-null bool
p2         2075 non-null object
p2_conf    2075 non-null float64
p2_dog     2075 non-null bool
p3         2075 non-null object
p3_conf    2075 non-null float64
p3_dog     2075 non-null bool
dtypes: bool(3), float64(3), int64(1), object(4)
memory usage: 152.0+ KB


In [24]:
# check if ids are unique
image_pred.index.is_unique

True

In [25]:
# check jpg_url
image_pred[~image_pred.jpg_url.str.endswith(('.jpg', '.png'), na = False)].jpg_url.count()

0

In [26]:
# check image number
image_pred.img_num.value_counts()

1    1780
2    198 
3    66  
4    31  
Name: img_num, dtype: int64

In [27]:
# check 1st prediction
image_pred.p1.sample(3)

tweet_id
694329668942569472    boxer        
701545186879471618    Border_collie
740995100998766593    malamute     
Name: p1, dtype: object

In [28]:
# check dog predictions
image_pred.p1_dog.count()

2075

#### Quality & Tidiness Issues

- the dataset has 2075 entries, while twitter archive dataset has 2356 entries;
- column names are confusing and do not give much information about the content;
- dog breeds contain underscores, and have different case formatting;
- only 2075 images have been classified as dog images for top prediction;
- dataset should be merged with the twitter archive dataset;

### Assess: Twitter API Data

In [29]:
# display sample of data
twitter_api.sample(3)

favorites  retweets
tweet_id                               
831262627380748289  13066      2350    
859074603037188101  35553      14740   
755206590534418437  18212      6148

In [30]:
# print a summary of a DataFrame
twitter_api.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2354 entries, 892420643555336193 to 666020888022790149
Data columns (total 2 columns):
favorites    2354 non-null int64
retweets     2354 non-null int64
dtypes: int64(2)
memory usage: 55.2 KB


In [31]:
# check if ids are unique
twitter_arch.index.is_unique

True

#### Quality & Tidiness Issues

- twitter archive dataset has 2356 entries, while twitter API data has 2354;
- dataset should be merged with the twitter archive dataset;

## Cleaning Data

Using pandas, clean the quality and tidiness issues identified in the [Assessing Data](#Assessing-Data) section.

<hr>

* [Twitter Archive Data](#Clean:-Twitter-Archive-Data)  
* [Image Predictions](#Clean:-Image-Predictions)  
* [Twitter API Data](#Clean:-Twitter-API-Data)
* [Merge Datasets](#Merge-Datasets)

### Clean: Twitter Archive Data

In [32]:
# create a copy of dataset
twitter_arch_clean = twitter_arch.copy()

In [33]:
# display sample of data
twitter_arch_clean.sample(3)

in_reply_to_status_id  in_reply_to_user_id  \
tweet_id                                                         
675710890956750848 NaN                    NaN                    
675489971617296384 NaN                    NaN                    
793962221541933056 NaN                    NaN                    

                                    timestamp  \
tweet_id                                        
675710890956750848  2015-12-12 16:16:45 +0000   
675489971617296384  2015-12-12 01:38:53 +0000   
793962221541933056  2016-11-02 23:45:19 +0000   

                                                                                                source  \
tweet_id                                                                                                 
675710890956750848  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
675489971617296384  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
793962221541933056  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                        text  \
tweet_id                                                                                                                                                       
675710890956750848  This is Lenny. He was just told that he couldn't explore the fish tank. 12/10 smh all that work for nothing https://t.co/JWi6YrpiO1        
675489971617296384  RT until we find this dog. Clearly a cool dog (front leg relaxed out window). Looks to be a superb driver. 10/10 https://t.co/MnTrKaQ8Wn   
793962221541933056  This is Maximus. His face is stuck like that. Tragic really. Great tongue tho. 12/10 would pet firmly https://t.co/xIfrsMNLBR              

                    retweeted_status_id  retweeted_status_user_id  \
tweet_id                                                            
675710890956750848 NaN                  NaN                         
675489971617296384 NaN                  NaN                         
793962221541933056 NaN                  NaN                         

                   retweeted_status_timestamp  \
tweet_id                                        
675710890956750848  NaN                         
675489971617296384  NaN                         
793962221541933056  NaN                         

                                                                                                                                      expanded_urls  \
tweet_id                                                                                                                                              
675710890956750848  https://twitter.com/dog_rates/status/675710890956750848/photo/1,https://twitter.com/dog_rates/status/675710890956750848/photo/1   
675489971617296384  https://twitter.com/dog_rates/status/675489971617296384/photo/1                                                                   
793962221541933056  https://twitter.com/dog_rates/status/793962221541933056/photo/1,https://twitter.com/dog_rates/status/793962221541933056/photo/1   

                    rating_numerator  rating_denominator     name  doggo  \
tweet_id                                                                   
675710890956750848  12                10                  Lenny    0       
675489971617296384  10                10                  None     0       
793962221541933056  12                10                  Maximus  0       

                    floofer  pupper  puppo  number_categories  
tweet_id                                                       
675710890956750848  0        0       0      0                  
675489971617296384  0        0       0      0                  
793962221541933056  0        0       0      0

#### Define

Some of the gathered tweets are replies and retweets
- remove replies data from the dataset
- remove retweets data from the dataset

#### Code

In [34]:
# display all columns
twitter_arch_clean.columns

Index(['in_reply_to_status_id', 'in_reply_to_user_id', 'timestamp', 'source',
       'text', 'retweeted_status_id', 'retweeted_status_user_id',
       'retweeted_status_timestamp', 'expanded_urls', 'rating_numerator',
       'rating_denominator', 'name', 'doggo', 'floofer', 'pupper', 'puppo',
       'number_categories'],
      dtype='object')

In [35]:
# drop unnecessary columns
twitter_arch_clean.drop(['in_reply_to_status_id','in_reply_to_user_id','retweeted_status_id',
           'retweeted_status_user_id','retweeted_status_timestamp'], axis = 1, inplace = True)

In [36]:
# display cleaned dataset
twitter_arch_clean.sample(3)

timestamp  \
tweet_id                                        
769335591808995329  2016-08-27 00:47:53 +0000   
670764103623966721  2015-11-29 00:39:59 +0000   
718460005985447936  2016-04-08 15:26:28 +0000   

                                                                                                source  \
tweet_id                                                                                                 
769335591808995329  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
670764103623966721  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   
718460005985447936  <a href="http://twitter.com/download/iphone" rel="nofollow">Twitter for iPhone</a>   

                                                                                                                                                           text  \
tweet_id                                                                                                                                                          
769335591808995329  RT @dog_rates: Ever seen a dog pet another dog? Both 13/10 truly an awe-inspiring scene. (Vid by @mdougherty20) https://t.co/3PoKf6cw7f       
670764103623966721  Meet Vincent. He's a wild Adderall Cayenne. Shipped for free. Always fresh. Never frozen. 10/10 great purchase https://t.co/ZfS7chSsi7        
718460005985447936  Meet Bowie. He's listening for underground squirrels. Smart af. Left eye is considerably magical. 9/10 would so pet https://t.co/JyNmyjy3fe   

                                                                      expanded_urls  \
tweet_id                                                                              
769335591808995329  https://vine.co/v/iXQAm5Lrgrh,https://vine.co/v/iXQAm5Lrgrh       
670764103623966721  https://twitter.com/dog_rates/status/670764103623966721/photo/1   
718460005985447936  https://twitter.com/dog_rates/status/718460005985447936/photo/1   

                    rating_numerator  rating_denominator     name  doggo  \
tweet_id                                                                   
769335591808995329  13                10                  None     0       
670764103623966721  10                10                  Vincent  0       
718460005985447936  9                 10                  Bowie    0       

                    floofer  pupper  puppo  number_categories  
tweet_id                                                       
769335591808995329  0        0       0      0                  
670764103623966721  0        0       0      0                  
718460005985447936  0        0       0      0

#### Define

The timestamp has an incorrect datatype - is an object, should be DateTime
* convert to datetime

#### Code

In [37]:
# convert to datetime
twitter_arch_clean.timestamp = pd.to_datetime(twitter_arch_clean.timestamp)

In [38]:
# display dataset types
twitter_arch_clean.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2356 entries, 892420643555336193 to 666020888022790149
Data columns (total 12 columns):
timestamp             2356 non-null datetime64[ns]
source                2356 non-null object
text                  2356 non-null object
expanded_urls         2297 non-null object
rating_numerator      2356 non-null int64
rating_denominator    2356 non-null int64
name                  2356 non-null object
doggo                 2356 non-null int64
floofer               2356 non-null int64
pupper                2356 non-null int64
puppo                 2356 non-null int64
number_categories     2356 non-null int64
dtypes: datetime64[ns](1), int64(7), object(4)
memory usage: 239.3+ KB


#### Define

Source is an HTML element - its text should be extracted
* extract inner text of the HTML elements

#### Code

In [39]:
# extract inner text from HTML
twitter_arch_clean.source = twitter_arch_clean.source.apply(lambda x: re.findall(r'>(.*)<', x)[0])

In [40]:
# display new source
twitter_arch_clean.source.value_counts()

Twitter for iPhone     2221
Vine - Make a Scene    91  
Twitter Web Client     33  
TweetDeck              11  
Name: source, dtype: int64

#### Define

Some rows in the text column begin from 'RT @dog_rates:'. Some rows have leading and/or trailing whitespace
- remove 'RT @dog_rates:'
- strip whitespace

#### Code

In [41]:
# example of tweet
twitter_arch_clean.text[838831947270979586]

"RT @dog_rates: This is Riley. His owner put a donut pillow around him and he loves it so much he won't let anyone take it off. 13/10 https:…"

In [42]:
# remove 'RT @dog_rates:' and strip leading and trailing space
twitter_arch_clean.text = twitter_arch_clean.text.str.replace('RT @dog_rates:', '').str.strip()

In [43]:
# example of tweet after clean up
twitter_arch_clean.text[838831947270979586]

"This is Riley. His owner put a donut pillow around him and he loves it so much he won't let anyone take it off. 13/10 https:…"

#### Define

We have 639 expanded urls which contain more than one url address and 59 missing expanded urls
- build correct links by using tweet id

#### Code

In [44]:
# fix expanded urls
for index, column in twitter_arch_clean.iterrows():
    twitter_arch_clean.loc[index, 'expanded_urls'] = 'https://twitter.com/dog_rates/status/' + str(index)

twitter_arch_clean.sample(3)

timestamp              source  \
tweet_id                                                     
861005113778896900 2017-05-06 23:49:50  Twitter for iPhone   
886366144734445568 2017-07-15 23:25:31  Twitter for iPhone   
856282028240666624 2017-04-23 23:01:59  Twitter for iPhone   

                                                                                                                                                                           text  \
tweet_id                                                                                                                                                                          
861005113778896900  This is Burt. He thinks your thesis statement is comically underdeveloped. 12/10 intellectual af https://t.co/jH6EN9cEn6                                      
886366144734445568  This is Roscoe. Another pupper fallen victim to spontaneous tongue ejections. Get the BlepiPen immediate. 12/10 deep breaths Roscoe https://t.co/RGE08MIJox   
856282028240666624  This is Cermet, Paesh, and Morple. They are absolute h*ckin superstars. Watered every day so they can grow. 14/10 for all https://t.co/GUefqUmZv8             

                                                              expanded_urls  \
tweet_id                                                                      
861005113778896900  https://twitter.com/dog_rates/status/861005113778896900   
886366144734445568  https://twitter.com/dog_rates/status/886366144734445568   
856282028240666624  https://twitter.com/dog_rates/status/856282028240666624   

                    rating_numerator  rating_denominator    name  doggo  \
tweet_id                                                                  
861005113778896900  12                10                  Burt    0       
886366144734445568  12                10                  Roscoe  0       
856282028240666624  14                10                  Cermet  0       

                    floofer  pupper  puppo  number_categories  
tweet_id                                                       
861005113778896900  0        0       0      0                  
886366144734445568  0        1       0      1                  
856282028240666624  0        0       0      0

#### Define

Float ratings have been incorrectly read from the text of tweet

- gather correct rating when rating is a fraction

#### Code

In [45]:
# example of tweet with incorrect rating
twitter_arch[twitter_arch.text.str.contains(r'\d+\.\d+\/\d+')][['text','rating_denominator', 'rating_numerator']].sample(1)

text  \
tweet_id                                                                                                                         
681340665377193984  I've been told there's a slight possibility he's checking his mirror. We'll bump to 9.5/10. Still a menace   

                    rating_denominator  rating_numerator  
tweet_id                                                  
681340665377193984  10                  5

In [46]:
# convert both columns to floats
twitter_arch_clean['rating_numerator'] = twitter_arch_clean['rating_numerator'].astype(float)
twitter_arch_clean['rating_denominator'] = twitter_arch_clean['rating_denominator'].astype(float)

# find columns with fractions
fraction_ratings = twitter_arch[twitter_arch.text.str.contains(r'\d+\.\d+\/\d+')].index

# extract correct rating and replace incorrect one
for index in fraction_ratings:
    rating = re.search('\d+\.\d+\/\d+', twitter_arch_clean.loc[index,:].text).group(0)
    twitter_arch_clean.at[index,'rating_numerator'], twitter_arch_clean.at[index,'rating_denominator'] = rating.split('/')

In [47]:
# display sample of fixed data
twitter_arch_clean.loc[fraction_ratings,:][['text','rating_denominator', 'rating_numerator']].sample(1)

text  \
tweet_id                                                                                                                                               
786709082849828864  This is Logan, the Chow who lived. He solemnly swears he's up to lots of good. H*ckin magical af 9.75/10 https://t.co/yBO5wuqaPS   

                    rating_denominator  rating_numerator  
tweet_id                                                  
786709082849828864  10.0                9.75

#### Define

Denominator of some ratings is not 10. Numerator of some ratings is greater than 10. 
- fix incorrectly read ratings
- normalize rating

#### Code

In [48]:
# save index of tweets with denominator greater than 10
high_denominator = twitter_arch[twitter_arch.rating_denominator > 10].index

# display sample of data with denominator greater than 10
twitter_arch_clean.loc[high_denominator,:][['text','rating_denominator', 'rating_numerator']].sample(1)

text  \
tweet_id                                                       
832088576586297345  @docmisterio account started on 11/15/15   

                    rating_denominator  rating_numerator  
tweet_id                                                  
832088576586297345  15.0                11.0

In [49]:
# fix rating manually for tweets for which rating was read incorrectly
twitter_arch_clean.loc[832088576586297345, 'rating_denominator'] = 0
twitter_arch_clean.loc[832088576586297345, 'rating_numerator'] = 0

twitter_arch_clean.loc[775096608509886464, 'rating_denominator'] = 10
twitter_arch_clean.loc[775096608509886464, 'rating_numerator'] = 14

twitter_arch_clean.loc[740373189193256964, 'rating_denominator'] = 10
twitter_arch_clean.loc[740373189193256964, 'rating_numerator'] = 14

twitter_arch_clean.loc[722974582966214656, 'rating_denominator'] = 10
twitter_arch_clean.loc[722974582966214656, 'rating_numerator'] = 13

twitter_arch_clean.loc[716439118184652801, 'rating_denominator'] = 10
twitter_arch_clean.loc[716439118184652801, 'rating_numerator'] = 11

twitter_arch_clean.loc[682962037429899265, 'rating_denominator'] = 10
twitter_arch_clean.loc[682962037429899265, 'rating_numerator'] = 10

In [50]:
# display sample of fixed rating
twitter_arch_clean.loc[high_denominator,:][['text','rating_denominator', 'rating_numerator']].sample(1)

text  \
tweet_id                                                                                                                                                           
740373189193256964  After so many requests, this is Bretagne. She was the last surviving 9/11 search dog, and our second ever 14/10. RIP https://t.co/XAVDNDaVgQ   

                    rating_denominator  rating_numerator  
tweet_id                                                  
740373189193256964  10.0                14.0

In [51]:
# normalize rating
twitter_arch_clean['rating'] = twitter_arch_clean['rating_numerator'] / twitter_arch_clean['rating_denominator']

In [52]:
# display sample of data with the new column
twitter_arch_clean[['text','rating_denominator', 'rating_numerator', 'rating']].sample(1)

text  \
tweet_id                                                                                                           
704499785726889984  When you wake up from a long nap and have no idea who you are. 12/10 https://t.co/dlF93GLnDc   

                    rating_denominator  rating_numerator  rating  
tweet_id                                                          
704499785726889984  10.0                12.0              1.2

#### Define

We have stop words in the name column and 'None' values
- replace stop words with the correct name
- replace None with Nan

#### Code

In [53]:
# adapted from https://stackoverflow.com/a/35976915/7382214

# keywords after which a dog name appears in the sentence
keywords = ('named', 'This is', 'Say hello to', 'Meet', 'Here we have', 'name is', 'name to')
# build regex pattern to find names
regex = r"\b(?:{})\b (\w+)".format("|".join(keywords))

# loop over the rows, and find dog names
for index, column in twitter_arch_clean.iterrows():
    split_list = re.split(regex, twitter_arch_clean.loc[index, 'text'], flags = re.IGNORECASE)
    if len(split_list) > 1 and split_list[1] not in stop_words:
        twitter_arch_clean.loc[index, 'name'] = split_list[1].title()
    else:
        twitter_arch_clean.loc[index, 'name'] = np.nan

#### Define

Dog 'stage' classification (doggo, floofer, pupper or puppo) should be one column. Some dogs have more than one category assigned

#### Code

In [54]:
twitter_arch_clean.head(2)

timestamp              source  \
tweet_id                                                     
892420643555336193 2017-08-01 16:23:56  Twitter for iPhone   
892177421306343426 2017-08-01 00:17:27  Twitter for iPhone   

                                                                                                                                                                                  text  \
tweet_id                                                                                                                                                                                 
892420643555336193  This is Phineas. He's a mystical boy. Only ever appears in the hole of a donut. 13/10 https://t.co/MgUWQ76dJU                                                        
892177421306343426  This is Tilly. She's just checking pup on you. Hopes you're doing ok. If not, she's available for pats, snugs, boops, the whole bit. 13/10 https://t.co/0Xxu71qeIV   

                                                              expanded_urls  \
tweet_id                                                                      
892420643555336193  https://twitter.com/dog_rates/status/892420643555336193   
892177421306343426  https://twitter.com/dog_rates/status/892177421306343426   

                    rating_numerator  rating_denominator     name  doggo  \
tweet_id                                                                   
892420643555336193  13.0              10.0                Phineas  0       
892177421306343426  13.0              10.0                Tilly    0       

                    floofer  pupper  puppo  number_categories  rating  
tweet_id                                                               
892420643555336193  0        0       0      0                  1.3     
892177421306343426  0        0       0      0                  1.3

In [55]:
# dog types
for index, column in twitter_arch_clean.iterrows():
    for word in ['puppo', 'pupper', 'doggo', 'floofer']:
        if word.lower() in twitter_arch_clean.loc[index, 'text'].lower():
            twitter_arch_clean.loc[index, 'dog_type'] = word
            
# drop old columns
twitter_arch_clean.drop(['puppo',
                       'pupper',
                       'doggo',
                       'floofer'],
                      axis=1, inplace=True)

In [56]:
# display sample of fixed data
twitter_arch_clean.sample(3)

timestamp              source  \
tweet_id                                                     
700002074055016451 2016-02-17 17:01:14  Twitter for iPhone   
869988702071779329 2017-05-31 18:47:24  Twitter for iPhone   
842846295480000512 2017-03-17 21:13:10  Twitter for iPhone   

                                                                                                                                                            text  \
tweet_id                                                                                                                                                           
700002074055016451  This is Thumas. He covered himself in nanners for maximum camouflage. It didn't work. I can still see u Thumas. 9/10 https://t.co/x0ZDlNqfb1   
869988702071779329  We only rate dogs. This is quite clearly a smol broken polar bear. We'd appreciate if you only send dogs. Thank you... 12/10…                  
842846295480000512  This is Charlie. He's wishing you a very fun and safe St. Pawtrick's Day. 13/10 festive af https://t.co/nFpNgCWWYs                             

                                                              expanded_urls  \
tweet_id                                                                      
700002074055016451  https://twitter.com/dog_rates/status/700002074055016451   
869988702071779329  https://twitter.com/dog_rates/status/869988702071779329   
842846295480000512  https://twitter.com/dog_rates/status/842846295480000512   

                    rating_numerator  rating_denominator     name  \
tweet_id                                                            
700002074055016451  9.0               10.0                Thumas    
869988702071779329  12.0              10.0                Quite     
842846295480000512  13.0              10.0                Charlie   

                    number_categories  rating dog_type  
tweet_id                                                
700002074055016451  0                  0.9     NaN      
869988702071779329  0                  1.2     NaN      
842846295480000512  0                  1.3     NaN

### Clean: Image Predictions

In [57]:
# create a copy of dataset
image_pred_clean = image_pred.copy()

In [58]:
# display sample of data
image_pred_clean.sample(3)

jpg_url  \
tweet_id                                                                                                      
783391753726550016  https://pbs.twimg.com/media/Ct8qn8EWIAAk9zP.jpg                                           
830956169170665475  https://pbs.twimg.com/ext_tw_video_thumb/830956118893543424/pu/img/t2G0raF7pDPRMAH5.jpg   
781524693396357120  https://pbs.twimg.com/media/CtiIj0AWcAEBDvw.jpg                                           

                    img_num                  p1   p1_conf  p1_dog  \
tweet_id                                                            
783391753726550016  4        Norwegian_elkhound  0.877130  True     
830956169170665475  1        kuvasz              0.451516  True     
781524693396357120  1        tennis_ball         0.994712  False    

                                          p2   p2_conf  p2_dog  \
tweet_id                                                         
783391753726550016  cairn                     0.086241  True     
830956169170665475  golden_retriever          0.317196  True     
781524693396357120  Chesapeake_Bay_retriever  0.003523  True     

                                    p3   p3_conf  p3_dog  
tweet_id                                                  
783391753726550016  keeshond            0.011019  True    
830956169170665475  English_setter      0.132759  True    
781524693396357120  Labrador_retriever  0.000921  True

#### Define

Column names are confusing and do not give much information about the content.  
- Change column names to more descriptive ones.

#### Code

In [59]:
# display current labels
image_pred_clean.columns

Index(['jpg_url', 'img_num', 'p1', 'p1_conf', 'p1_dog', 'p2', 'p2_conf',
       'p2_dog', 'p3', 'p3_conf', 'p3_dog'],
      dtype='object')

In [60]:
# change labels
image_pred_clean.columns = ['image_url', 
                            'img_number', 
                            '1st_prediction',
                            '1st_prediction_confidence',
                            '1st_prediction_isdog',
                            '2nd_prediction',
                            '2nd_prediction_confidence',
                            '2nd_prediction_isdog',
                            '3rd_prediction',
                            '3rd_prediction_confidence',
                            '3rd_prediction_isdog']

In [61]:
# display new labels
image_pred_clean.columns

Index(['image_url', 'img_number', '1st_prediction',
       '1st_prediction_confidence', '1st_prediction_isdog', '2nd_prediction',
       '2nd_prediction_confidence', '2nd_prediction_isdog', '3rd_prediction',
       '3rd_prediction_confidence', '3rd_prediction_isdog'],
      dtype='object')

#### Define

Dog breeds contain underscores, and have different case formatting
- Replace underscores with whitespace
- Capitalize the first letter of each word

#### Code

In [62]:
# columns with dog breed
dog_breed_cols = ['1st_prediction', '2nd_prediction', '3rd_prediction']

# remove underscore and capitalize the first letter of each word 
for column in dog_breed_cols:
    image_pred_clean[column] = image_pred_clean[column].str.replace('_', ' ').str.title()

In [63]:
# display sample of changes
image_pred_clean[dog_breed_cols].sample(3)

1st_prediction       2nd_prediction  \
tweet_id                                                 
673363615379013632  Ox             Warthog               
743980027717509120  Bull Mastiff   Rhodesian Ridgeback   
677918531514703872  Eskimo Dog     Dalmatian             

                                    3rd_prediction  
tweet_id                                            
673363615379013632  Bison                           
743980027717509120  Pug                             
677918531514703872  American Staffordshire Terrier

#### Define

Only 2075 images have been classified as dog images for top prediction
- If 1st predictions is not a dog breed, then use dog breed predicted in the 2nd or 3rd predicion

#### Code

In [64]:
# build function to determine dog breed
# if no breed detected, set value to NaN
# adapted from https://stackoverflow.com/a/26887820/7382214

def get_breed(row):
    if row['1st_prediction_isdog'] == True:
        return row['1st_prediction'], row['1st_prediction_confidence']
    if row['2nd_prediction_isdog'] == True:
        return row['2nd_prediction'], row['2nd_prediction_confidence']
    if row['3rd_prediction_isdog'] == True:
        return row['3rd_prediction'], row['3rd_prediction_confidence']
    return np.nan, np.nan

# apply function to dataset
# create new columns with data
image_pred_clean[['breed_predicted', 'prediction_confidence']] = pd.DataFrame(image_pred_clean.apply(lambda row: get_breed(row), axis = 1).tolist(), index = image_pred_clean.index) 

# drop old columns
image_pred_clean.drop(['1st_prediction',
                       '1st_prediction_confidence',
                       '1st_prediction_isdog',
                       '2nd_prediction',
                       '2nd_prediction_confidence',
                       '2nd_prediction_isdog',
                       '3rd_prediction',
                       '3rd_prediction_confidence',
                       '3rd_prediction_isdog'],
                      axis=1, inplace=True)

# drop rows without dog breed prediction
image_pred_clean.dropna(subset = ['breed_predicted', 'prediction_confidence'], inplace = True)

In [65]:
# displa sample of cleaned dataset
image_pred_clean.sample(3)

image_url  \
tweet_id                                                              
672622327801233409  https://pbs.twimg.com/media/CVWicBbUYAIomjC.jpg   
753294487569522689  https://pbs.twimg.com/media/CnQ9Vq1WEAEYP01.jpg   
699434518667751424  https://pbs.twimg.com/media/CbTj--1XEAIZjc_.jpg   

                    img_number   breed_predicted  prediction_confidence  
tweet_id                                                                 
672622327801233409  1           Golden Retriever  0.952773               
753294487569522689  1           Chow              0.194773               
699434518667751424  1           Golden Retriever  0.836572

This concludes cleaning activities for the Image Predictions dataset. The remaining task is to merge it with the Twitter archive data, which is covered in the [Clean: Merge Datasets](#Clean:-Merge-Datasets) section.

### Clean: Twitter API Data

In [66]:
# display sample of data
twitter_api.sample(3)

favorites  retweets
tweet_id                               
819952236453363712  5927       1369    
724046343203856385  2901       620     
691459709405118465  4449       1294

There is no need to perform cleaning tasks in this data set, except for merging it with the Twitter archive data, which is covered in the next section.

### Clean: Merge Datasets

## Analyze Data

Analyze and visualize data using matplotlib.

<hr>